ballerina/vector-store.bal

```diff
@@ -89,17 +89,55 @@ public distinct isolated class InMemoryVectorStore {
     # + return - An array of vector matches sorted by similarity score (limited to topK), 
     # or an `ai:Error` if the query fails
     public isolated function query(VectorStoreQuery query) returns VectorMatch[]|Error {
-        if query.embedding !is Vector {
-            return error Error("InMemoryVectorStore supports dense vectors exclusively");
-        }
-
         lock {
-            VectorMatch[] sorted = from var entry in self.entries
+            VectorStoreQuery clonedQuery = query.cloneReadOnly();
+            Embedding? embedding = clonedQuery.embedding;
+            MetadataFilters? filters = clonedQuery.filters;
+            if embedding !is Vector? {
+                return error Error("InMemoryVectorStore supports dense vectors exclusively");
+            }
+            if embedding is () {
+                readonly & VectorMatch[] results;
+                if filters is () {
+                    results = from InMemoryVectorEntry entry in self.entries
+                    select {
+                        chunk: entry.chunk.cloneReadOnly(),
+                        embedding: entry.embedding.cloneReadOnly(),
+                        similarityScore: 1.0,
+                        id: entry.id
+                    };
+                    return results.cloneReadOnly();
+                }
+                MetadataFilters metadataFilters = filters.cloneReadOnly();
+                results = from InMemoryVectorEntry entry in self.entries
+                    where check entryMatchesFilters(entry, metadataFilters)
+                    select {
+                        chunk: entry.chunk.cloneReadOnly(),
+                        embedding: entry.embedding.cloneReadOnly(),
+                        similarityScore: 1.0,
+                        id: entry.id
+                    };
+                return results.cloneReadOnly();
+            }
+            VectorMatch[] results = from InMemoryVectorEntry entry in self.entries
```

Comment:
I think, At this level we first need to apply the metadata filter first and do the similarity search. Can you double check this?